[HUDI-1075] Implement simple clustering strategies to create and run ClusteringPlan

satishkotha opened the pull request:
## What is the purpose of the pull request

 Implement simple clustering strategies to create and run ClusteringPlan

## Brief change log
- Add simple strategy to to schedule clustering and create ClusteringPlan. Initial strategy can limit partitions and bound the amount of data considered for clustering
- Add simple strategy to run the clustering plan and create 'replacecommit'. Initial strategy is implemented by using BulkInsert partitioner. 
- This can work for both COW and MOR tables. Clustering also does 'compaction' if there are log files. We ensure compaction and clustering do not work on same file groups.
- Added basic support for sorting by custom column(s). I may refactor this little bit. 
- Few other things to pay attention:

1.  After clustering all new records have a new commit time. I'm trying to see if it's possible to preserve commit_time from original file to support incremental queries.
2. Schema is not serializable. So I pass around schema string in some cases. Let me know if there is a better way to restructure code.
3. Initial strategy is implemented based on requirements for Uber. But the interfaces are flexible to allow other implementation. But please double check if I made bad assumptions in any of the strategies.

## Verify this pull request
- Manually tested 'inline clustering' end-to-end in local docker setup.
- I will be working on adding lot more unit tests (Publishing this PR to get high level feedback).
- Plan to add integ-tests using test-suite as part of separate PR.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java:
```diff
@@ -726,6 +751,54 @@ private void rollbackPendingCommits() {
+    return clusteringInstantOpt;
+  }
+
+  public void rollbackInflightClustering(HoodieInstant inflightInstant, HoodieTable<T, I, K, O> table) {
```
Should this be a public method ? Rolling back inflight could probably remain private ?

satishkotha replied:
Changed to protected. I dont think this is needed for CLI.  We can change it back to public, when needed.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,155 @@
+public class HoodieClusteringConfig extends DefaultHoodieConfig {
+
+  public static final String SCHEDULE_CLUSTERING_STRATEGY_CLASS = "hoodie.clustering.schedule.strategy.class";
+  public static final String DEFAULT_SCHEDULE_CLUSTERING_STRATEGY_CLASS =
```
Looks like the execution framework details are in the configs. Would be nice to have a wrapper around reflection utils that can pick the right execution engine based class but keep the strategy class name generic...not sure if it's too late for that..

satishkotha replied:
This is just the class name. It is possible to support both flink and spark clients with same class. Initial implementation is primarily targeted for Spark. So i explicitly called the class Spark.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,155 @@
+  public static final String DEFAULT_RUN_CLUSTERING_STRATEGY_CLASS =
+      "org.apache.hudi.client.clustering.run.SparkBulkInsertBasedRunClusteringStrategy";
+
+  // Turn on inline clustering - after few commits a inline compaction will be run
```
extreme nit : s/a/an :)

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,155 @@
+      return this;
+    }
+
+    public Builder withClusteringMaxGroupSize(long clusteringMaxGroupSize) {
```
Is this in MB ? Can we rename this to withClusteringMaxGroupSizeInMB ?

satishkotha replied:
Fixed

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,155 @@
+      return this;
+    }
+
+    public Builder withClusteringTargetFileSize(long targetFileSize) {
```
Same this, if in MB, add it to the name ?

satishkotha replied:
Fixed

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/HoodieTable.java:
```diff
@@ -326,6 +327,28 @@ public HoodieActiveTimeline getActiveTimeline() {
+   * @param context HoodieEngineContext
+   * @param clusteringInstantTime Instant Time
+   */
+  public abstract HoodieWriteMetadata<O> clustering(HoodieEngineContext context,
```
Should this be named "cluster" given the write client also has "cluster" ?

satishkotha replied:
Fixed

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/PartitionAwareScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,108 @@
+   */
+  protected abstract Stream<HoodieClusteringGroup> buildClusteringGroupsForPartition(String partitionPath,
+                                                                                     List<FileSlice> fileSlices,
+                                                                                     long maxDataInGroup);
```
Better name for "maxDataInGroup" ?

satishkotha replied:
Removed this argument per Li Wei suggestion.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/RunClusteringStrategy.java:
```diff
@@ -0,0 +1,67 @@
+   * file groups created is bounded by numOutputGroups.
+   * Note that commit is not done as part of strategy. commit is callers responsibility.
+   */
+  public abstract O performClustering(final I inputRecords, final int numOutputGroups, final String instantTime,
```
Can you expand on why we need the "schema" parameter ?

satishkotha replied:
This is "readerSchema" for all inputRecords being clustered. This is useful to read values for specified sort columns  from inputRecords.  Let me know if theres a better way to do this.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,179 @@
+    return slices.stream().map(slice -> new HoodieSliceInfo().newBuilder()
+        .setPartitionPath(slice.getPartitionPath())
+        .setFileId(slice.getFileId())
+        .setDataFilePath(slice.getBaseFile().map(BaseFile::getPath).orElse(""))
```
Could we use StringUtils.EMPTY or create a constant ?

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,179 @@
+/**
+ * Pluggable implementation for scheduling clustering and creating ClusteringPlan.
+ */
+public abstract class ScheduleClusteringStrategy<T extends HoodieRecordPayload,I,K,O> implements Serializable {
```
Could we split this -> https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/strategy/CompactionStrategy.java into IStrategy that should have common methods expected from all strategies such as buildMetrics, getPlanVersion..and then ScheduleClusteringStrategy can extend it ?
After that, let's move the buildMetrics into a utils class given it's the same across both compaction and clustering right now ?

satishkotha replied:
I feel like inheritance is wrong pattern here. These look similar now, but i think metrics will change differently over time. So having tight integration may make it difficult for these two evolve differently. Let me know if you have strong opinion. I can change it.

n3nash replied:
My general idea was to consolidate different kinds of strategies for operations such as compaction, clustering right now (tiering etc in the future). But I think it does take away some flexibility. I'm OK to keep it different right now.

satishkotha replied:
I moved buildMetrics to utils class. If you have strong opinion on common interface I can add it.

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/AbstractBulkInsertHelper.java:
```diff
@@ -27,8 +27,20 @@
 public abstract class AbstractBulkInsertHelper<T extends HoodieRecordPayload, I, K, O, R> {
 
+  /**
+   * Mark instant as inflight, write input records, updateIndex and return result.
```
nit : updateIndex -> update index, lets follow one pattern

satishkotha replied:
Fixed

n3nash commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/RunClusteringStrategy.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Pluggable implementation for writing data into new file groups based on ClusteringPlan.
+ */
+public abstract class RunClusteringStrategy<T extends HoodieRecordPayload,I,K,O> implements Serializable {
```
nit : s/RunClusteringStrategy/ExecuteClusteringStrategy ?

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -296,6 +298,57 @@ protected void completeCompaction(HoodieCommitMetadata metadata, JavaRDD<WriteSt
+      long durationInMs = metrics.getDurationInMs(clusteringTimer.stop());
+      try {
+        metrics.updateCommitMetrics(HoodieActiveTimeline.COMMIT_FORMATTER.parse(clusteringCommitTime).getTime(),
+            durationInMs, metadata, HoodieActiveTimeline.COMPACTION_ACTION);
```
Should this be a REPLACE_ACTION instead of COMPACTION_ACTION ?

satishkotha replied:
Good catch, fixed.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/SparkBulkInsertBasedRunClusteringStrategy.java:
```diff
@@ -0,0 +1,81 @@
+ * 1) Spark execution engine.
+ * 2) Uses bulk_insert to write data into new files.
+ */
+public class SparkBulkInsertBasedRunClusteringStrategy<T extends HoodieRecordPayload<T>>
```
Why is this not under the strategy package ?

satishkotha replied:
Moved.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/schedule/SparkBoundedDayBasedScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,107 @@
+ * 1) Spark execution engine.
+ * 2) Limits amount of data per clustering operation.
+ */
+public class SparkBoundedDayBasedScheduleClusteringStrategy<T extends HoodieRecordPayload<T>>
```
same question, should this go under strategy package ?

satishkotha replied:
Moved

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/SparkLazyInsertIterable.java:
```diff
@@ -34,14 +35,18 @@
 
 public class SparkLazyInsertIterable<T extends HoodieRecordPayload> extends HoodieLazyInsertIterable<T> {
 
+  private boolean addMetadataFields;
```
Could you explain this change please ?

satishkotha replied:
With regular bulk insert, RDD only has user specified columns. With bulk insert based clustering, RDD also has hoodie internal columns. So, I am adding this flag to make it work for both cases. Let me know if you want me to reorganize this  differently.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/BulkInsertMapFunction.java:
```diff
@@ -41,20 +41,22 @@
   public BulkInsertMapFunction(String instantTime, boolean areRecordsSorted,
                                HoodieWriteConfig config, HoodieTable hoodieTable,
-                               List<String> fileIDPrefixes) {
+                               List<String> fileIDPrefixes, boolean addMetadataFields) {
```
same

satishkotha replied:
With regular bulk insert, RDD only has user specified columns. With bulk insert based clustering, RDD also has hoodie internal columns. So, I am adding this flag to make it work for both cases. Let me know if you want me to reorganize this  differently.

n3nash replied:
This problem exists today with the MOR tables when inserts go into log files. For this reason, we have a different constructor for the create handle that is called from the compaction code path -> https://github.com/apache/hudi/blob/36ce5bcd92fa6cb2fb88677eae1a7a846f606e3f/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCreateHandle.java#L93. Can you try to do the same for clustering and avoid passing this flag here ?

satishkotha replied:
We need this in transform function below (We dont use HoodieCreateHandle here). I changed the name to follow same pattern as HoodieCreateHandle to make it more clear. Let me know if this is clear.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,153 @@
+
+    JavaSparkContext engineContext = HoodieSparkEngineContext.getSparkContext(context);
+    // read rdd from input groups in plan
+    JavaRDD<WriteStatus> writeStatuses = clusteringPlan.getInputGroups().stream()
```
This runs the clustering groups in sequence, is that what is expected ?

satishkotha replied:
I changed it to parallelize using java thread pool. PTAL.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,153 @@
+    HoodieInstant instant = HoodieTimeline.getReplaceCommitRequestedInstant(instantTime);
+    // Mark instant as clustering inflight
+    table.getActiveTimeline().transitionReplaceRequestedToInflight(instant, Option.empty());
+    table.getMetaClient().reloadActiveTimeline();
```
Why do we need this reload ?

satishkotha replied:
After this step, we convert inflight to commit. Without this reload, 'inflight' doesnt exist in table.getActiveTimeline(). So commit doesnt succeed. We seem to be doing similar step for compaction too.

n3nash replied:
Okay

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieFileSliceReader.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Collection;
+import java.util.Iterator;
+
+public class HoodieFileSliceReader implements Iterable<HoodieRecord<? extends HoodieRecordPayload>> {
```
Java doc please

satishkotha replied:
Added

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkScheduleClusteringActionExecutor.java:
```diff
@@ -0,0 +1,74 @@
+    int commitsSinceLastClustering = table.getActiveTimeline().getCommitsTimeline().filterCompletedInstants()
+        .findInstantsAfter(lastClusteringInstant.map(HoodieInstant::getTimestamp).orElse("0"), Integer.MAX_VALUE)
+        .countInstants();
+    if (config.getInlineClusterMaxCommits() > commitsSinceLastClustering) {
```
At a high level, what is the use-case when we would need to cluster every N commits ? For compaction it makes sense..

satishkotha replied:
I think this still makes sense for clustering too. Imagine, we are writing new data every 5minutes. It doesnt make sense to do file listing and figure out if theres enough data for clustering.  User knows there is enough data after N minutes, so they can select N/5 commits here. 

We can also potentially move this to strategy specific implementation. But I do think most strategies will use this. Let me know if you want me to move this into strategy.

n3nash replied:
I think this is one more config for people to manage. Clustering use-case spans across many different angles, compacting small files into larger being just one of them. For compaction, since there aren't other use-cases, every N commit makes total sense. But the other side is to keep the same behaviors consistent across compaction and clustering so keeping this would provide users the same knobs. 
I'll let @vinothchandar take a call on this.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,153 @@
+    return writeMetadata;
+  }
+
+  private JavaRDD<WriteStatus> runClusteringForGroup(HoodieClusteringGroup clusteringGroup, Map<String, String> strategyParams) {
```
So if there are N clustering groups consisting of M file groups that are being converted into O file groups, we will have N different RDD's. There are some performance limitation of unioning multiple different RDDs from what I remember. Can you run a simple test a) take 10000 records, parallelize them with jsc to create 1 RDD and then write them out to a file b) take 10000 records, create 1 RDD for each record and then union them and write them out to a single file ?

satishkotha replied:
Is there an alternative to union? Should we collect individual RDD and merge lists? I did testing only with 2 groups so far. So, I didn't see big performance degradation. I'll look into testing large number of groups as separate task. (We may also don't want number of groups to be really high as we want clustering to be atomic operation. Doing large amount of data/groups can increase chance of failures and slow down entire process.)

n3nash replied:
Yes, like we discussed offline, we should have a config to limit the number of clustering groups to avoid someone setting wrong configs for things like maxDataPerGroup and then result in a large number of RDDs to be unioned. I think we should still do a quick test to see if there is any degradation due to union before we land this diff so we are aware. It should be as simple as writing a test case that loops over code

TestUnionPerformance

long startTime = System.currentMillis();
RDD rdd = jsc.parallelize(Arrays.asList(record))
for (record in records)
{
rdd.union(jsc.parallelize(Arrays.asList(record)))
}
rdd.collect();
writeClient.bulkInsert(rdd);
sout("time taken with union" + System.currentMillis - starttime)

long startTime = System.currentMillis();
RDD rdd = jsc.parallelize(records)
rdd.collect();
writeClient.bulkInsert(rdd);
sout("time taken without union" + System.currentMillis - starttime)

satishkotha replied:
Ok, Changed to limit max groups to 50 (thats where we see significant degradation). I only applied for this strategy (because reduce is inside strategy code).

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,153 @@
+      LOG.info("MaxMemoryPerCompaction run as part of clustering => " + maxMemoryPerCompaction);
+      try {
+        Schema readerSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getSchema()));
+        HoodieFileReader<? extends IndexedRecord> baseFileReader = HoodieFileReaderFactory.getFileReader(table.getHadoopConf(), new Path(sliceInfo.getDataFilePath()));
```
Do we need this level of file readers in this part of the code ? Can we just call compactor.compact(context, compactionPlan, table, config, instantTime) ?

satishkotha replied:
compactor.compact seems to write data into new files and returns WriteStatus. We then have to read data and use it for clustering. I think this additional is unnecessary.  Also, creating compactionplan for clustering looks like tight coupling to me. This is only few lines of code and abstractions look reasonable to me. Please take a look and let me know if you have any suggestions.

n3nash reviewed (changes requested):
@satishkotha left some comments

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,155 @@
+  private static final String DEFAULT_INLINE_CLUSTERING_NUM_COMMITS = "4";
+
+  public static final String CLUSTERING_TARGET_PARTITIONS = "hoodie.clustering.target.partitions";
+  public static final String DEFAULT_CLUSTERING_TARGET_PARTITIONS = String.valueOf(2);
```
hello , i am interesting  about the scenario to limit the partition num of clustering. does cluster all partitions have some problem?

satishkotha replied:
I think we do file listing per partition in the first strategy implementation. So if you have lot of partitions, that can cause some degradation. I reached out on slack to help with testing. I sent you commands to run inline clustering. Will you be able to test this PR by setting this config to a large value?

lw309637554 replied:
1. "So if you have lot of partitions, that can cause some degradation" will resolved in file listing  https://issues.apache.org/jira/browse/HUDI-1292?
2. i am happy to test this scenario

satishkotha replied:
Yes, that ticket can resolve issues to some extent. But, in some cases, there are large day based tables with 1000s of partitions. we probably dont want to scan all partitions in those cases. So i think its useful to have this config.

satishkotha replied:
@lw309637554 let me know if you got a chance to test this scenario (or clustering in general)

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/PartitionAwareScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,108 @@
+   */
+  protected abstract Stream<HoodieClusteringGroup> buildClusteringGroupsForPartition(String partitionPath,
+                                                                                     List<FileSlice> fileSlices,
+                                                                                     long maxDataInGroup);
```
The implement of buildClusteringGroupsForPartition method in SparkBoundedDayBasedScheduleClusteringStrategy. Get params about clustering from maxDataInGroup、getWriteConfig().getClusteringTargetFileSize()、getWriteConfig().getClusteringTargetFileSize(). Maybe other implement will need more params . Can we just abstract buildClusteringGroupsForPartition with the HoodieWriteConfig for param ?  maybe like :
protected abstract Stream<HoodieClusteringGroup> buildClusteringGroupsForPartition(String partitionPath,
                                                                                     List<FileSlice> fileSlices,
                                                                                     HoodieWriteConfig writeConfig);

satishkotha replied:
Sure. Changed it to keep it more flexible.

lw309637554 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/SparkBulkInsertBasedRunClusteringStrategy.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.hudi.config.HoodieClusteringConfig.SORT_COLUMNS_PROPERTY;
```
if use HoodieClusteringConfig.SORT_COLUMNS_PROPERTY in code  instead of import static better?

satishkotha replied:
for constants, we seem to be using static imports in many places. So I keep as is.

lw309637554 commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDCustomColumnsSortPartitioner.java:
```diff
@@ -0,0 +1,66 @@
+  @Override
+  public JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records,
+                                                     int outputSparkPartitions) {
+    final String[] sortColumns = this.sortColumnNames;
```
if we should  constraint the columns start of partitionpath and row_key. Or just auto fill the two columns? Because the records of same partition and row_key need in one file slice.

satishkotha replied:
I dont think we want to add this constraint. There are use cases where we want to sort data by other columns.

lw309637554 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,179 @@
+    Set<HoodieFileGroupId> fgIdsInPendingCompactionAndClustering = ((SyncableFileSystemView) getHoodieTable().getSliceView()).getPendingCompactionOperations()
+        .map(instantTimeOpPair -> instantTimeOpPair.getValue().getFileGroupId())
+        .collect(Collectors.toSet());
+    fgIdsInPendingCompactionAndClustering.addAll(ClusteringUtils.getAllFileGroupsInPendingClusteringPlans(getHoodieTable().getMetaClient()).keySet());
```
can we get pendingClusterFileGroupid  from getHoodieTable().getSliceView() just like getPendingCompactionOperations?

satishkotha replied:
Good catch. I changed it to get this data from view.

satishkotha commented:
@n3nash Please take another look.

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDCustomColumnsSortPartitioner.java:
```diff
@@ -0,0 +1,66 @@
+                                                     int outputSparkPartitions) {
+    final String[] sortColumns = this.sortColumnNames;
+    final String schemaStr = this.schemaString;
+    return records.sortBy(record -> {
```
I have a suggestion to change the way we are generating the sort cols. Instead of performing this operation during write where the following code will spend cycles to convert data to generic record again, can we do this during read time when the HoodieRecord is being constructed ? I think you are using HoodieFileSliceReader and HoodieMergedRecordScanner to do this, does it make sense to add a sortKey to HoodieRecord and when you are constructing the hoodie record where you anyways have a handle to the generic record, you can just set that value in the hoodie record ? 
The downside of this approach is that the RDD bloats up so the shuffle will be larger so the question is whether that's worse or converting to generic record is worse from a CPU perspective

satishkotha replied:
I like this suggestion. This 'sortColumn' is probably needed only for this 'BulkInsert' based clustering strategy.  So, just want to make sure its ok to add new fields in HoodieRecord.  If you think thats acceptable, I'll make the change. Let me know.

n3nash replied:
@satishkotha Do we know the overhead of converting schemaString to schema and generic record ?

satishkotha replied:
@n3nash I dont have numbers for this. I'll add metrics and calculate this. If you have any details from past examples, let me know

n3nash commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,159 @@
+        try {
+          Schema readerSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getSchema()));
+          HoodieFileReader<? extends IndexedRecord> baseFileReader = HoodieFileReaderFactory.getFileReader(table.getHadoopConf(), new Path(sliceInfo.getDataFilePath()));
+          HoodieMergedLogRecordScanner scanner = new HoodieMergedLogRecordScanner(table.getMetaClient().getFs(),
```
The merged record scanner will try to keep an in-memory map, can we optimize for cases when there are no log files ? Or am I missing where this is done for COW / MOR where we are only reading parquet files ?

satishkotha replied:
I initially had implemented separate for COW/MOR. But it did not result in significant difference in performance in terms of run time. Memory utilization is definitely more with log files. But since we have memory limit configuration, to keep it simple, I just added one path that works for both COW/MOR.

I brought back earlier implementation. Take a look. If you think thats not complex, we can keep both.

n3nash replied:
@satishkotha I see the new one looks messy, if the runtime/difference in performance is the same for MergedScanner and no memory overhead of FileSlices without any log files, let's keep your other implementation with MergedScanner for both. Added 1 comment on the FileSliceReader

n3nash reviewed (changes requested):
@satishkotha left some more comments, PTAL

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieFileSliceReader.java:
```diff
@@ -0,0 +1,62 @@
+  private HoodieMergedLogRecordScanner logRecordScanner;
+
+  public static <R extends IndexedRecord, T extends HoodieRecordPayload> HoodieFileSliceReader getFileSliceReader(
+      HoodieFileReader<R> baseFileReader, HoodieMergedLogRecordScanner scanner, Schema schema, String payloadClass) throws IOException {
```
Should we make this AbstractLogRecordScanner to allow for UnMergedScanner as well ? @satishkotha

satishkotha replied:
UnmergedScanner doesnt seem to have a way to get record iterator. This requires lot more refactoring if we want to make it work for both.

n3nash replied:
Hmm, okay, in that case, can we keep the method signature of FileSliceReader as iterator rather than scanner so that way the FileSliceReader is generic and then we can figure out how to make those record scanner changes later ?
@satishkotha

satishkotha replied:
We also need process method in LogRecordScanner. So I kept this 'factory' method as is. But changed HoodieFileSliceReader constructor to take records. We can implement another factory method for UnmergedScanner. PTAL.

n3nash replied:
@satishkotha We cannot use the scanner.getRecords() API, that will bring all the records into memory and cause the job to OOM. Let's just continue to return the iterator and then lazily fetch records as you iterate.

satishkotha replied:
@n3nash getRecords just retruns 'ExternalSpillableMap'. So i dont think it can cause OOM. Anyway, changed it to iterator to keep it similar to reading records from base files.

vinothchandar replied:
Nonetheless we should file a `code cleanup` JIRA to provide these iterators as core building blocks under a nice abstractions.

n3nash reviewed (changes requested):
@satishkotha high level 2 outstanding questions 
a) Moving sortKey into HoodieRecord is needed only if we see performance degradation in conversion, if not we can keep it that way
b) Keep the old way of reading file slice with and without log files as long as no memory overhead for without log files

satishkotha commented:
> @satishkotha high level 2 outstanding questions
> a) Moving sortKey into HoodieRecord is needed only if we see performance degradation in conversion, if not we can keep it that way
> b) Keep the old way of reading file slice with and without log files as long as no memory overhead for without log files

@n3nash Both of these are performance related. My guess is this takes 3-4 weeks to test out with proper setup at scale (there are some issues because of separate branches etc). I think I addressed all of your other comments and fixed few issues I found with small scale.

 Let me know if you have any other feedback.

n3nash commented:
@satishkotha Left 1 more comment around FileSliceReader. For the 2 performance related comments, can you please create those tickets so we can track that and improve as we do scale testing. Once these are done, I can approve the diff.

n3nash commented:
@satishkotha Please create the follow up tickets for performance and I will approve this PR.
@vinothchandar Please take your pass on the PR

satishkotha commented:
@n3nash  created https://issues.apache.org/jira/browse/HUDI-1443 and https://issues.apache.org/jira/browse/HUDI-1442 for performance related measurements

n3nash reviewed (approved):
LGTM

lw309637554 commented:
> > @satishkotha high level 2 outstanding questions
> > a) Moving sortKey into HoodieRecord is needed only if we see performance degradation in conversion, if not we can keep it that way
> > b) Keep the old way of reading file slice with and without log files as long as no memory overhead for without log files
> 
> @n3nash Both of these are performance related. My guess is this takes 3-4 weeks to test out with proper setup at scale (there are some issues because of separate branches etc). I think I addressed all of your other comments and fixed few issues I found with small scale.
> 
> Let me know if you have any other feedback.

@satishkotha i also plan to do some performance test, such as kafka message large-scale  insert and then clustering period.

vinothchandar commented:
@satishkotha 

>After clustering all new records have a new commit time. I'm trying to see if it's possible to preserve commit_time from original file to support incremental queries.

this would be good to do. Can we file a tracking JIRA for this?  we can keep under the same umbrella with 0.8.0 fix version?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBulkInsertHelper.java:
```diff
@@ -59,25 +58,39 @@ public static SparkBulkInsertHelper newInstance() {
-                                                              BaseCommitActionExecutor<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, R> executor,
-                                                              boolean performDedupe,
-                                                              Option<BulkInsertPartitioner<T>> userDefinedBulkInsertPartitioner) {
+  public HoodieWriteMetadata<JavaRDD<WriteStatus>> bulkInsert(final JavaRDD<HoodieRecord<T>> inputRecords, final String instantTime, final HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> table, final HoodieWriteConfig config, final BaseCommitActionExecutor<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, R> executor, final boolean performDe ...
```
place each parameter on its own line?

satishkotha replied:
Done

vinothchandar commented on hudi-common/src/main/avro/HoodieClusteringGroup.avsc:
```diff
@@ -40,6 +40,11 @@
          "default": null
       },
+      {
+         "name":"numOutputGroups",
```
should it be called `numOutputSlices`?  the whole thing  is called a group right?  or may be you mean `numOutputFileGroups`? in which case, lets rename this to clarify

satishkotha replied:
Changed it to numOutputFileGroups

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/FileSliceUtils.java:
```diff
@@ -0,0 +1,67 @@
+  public static final String TOTAL_LOG_FILE_SIZE = "TOTAL_LOG_FILES_SIZE";
+  public static final String TOTAL_LOG_FILES = "TOTAL_LOG_FILES";
+
+  public static void addFileSliceCommonMetrics(List<FileSlice> fileSlices, Map<String, Double> metrics, long defaultBaseFileSize) {
```
this is very metric specific. could we stick it somewhere else closer to actual usage in hudi-client-common? may be rename the class to something like `FileSliceMetricUtils` ?

satishkotha replied:
Moved it as suggested.

vinothchandar reviewed (commented):
This is great, thanks @satishkotha ! 

I have completed a first pass. Don't have major concerns. May be we can work through these initial comments, as I complete the remainder. 

On follow ups
- IIUC inline clustering should work as-is from datasource/deltastreamer paths with this change, by passing necessary configs. We should create two JIRA one each for support for async clustering via datasource and deltastreamer?
- Can you share how much testing on a production environment has been done for this.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,165 @@
+  public static final String INLINE_CLUSTERING_PROP = "hoodie.clustering.inline";
+  private static final String DEFAULT_INLINE_CLUSTERING = "false";
+
+  public static final String INLINE_CLUSTERING_NUM_COMMIT_PROP = "hoodie.clustering.inline.num.commits";
```
Added

satishkotha commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBulkInsertHelper.java:
```diff
@@ -59,25 +58,39 @@ public static SparkBulkInsertHelper newInstance() {
-                                                              BaseCommitActionExecutor<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, R> executor,
-                                                              boolean performDedupe,
-                                                              Option<BulkInsertPartitioner<T>> userDefinedBulkInsertPartitioner) {
+  public HoodieWriteMetadata<JavaRDD<WriteStatus>> bulkInsert(final JavaRDD<HoodieRecord<T>> inputRecords, final String instantTime, final HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> table, final HoodieWriteConfig config, final BaseCommitActionExecutor<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>, R> executor, final boolean performDe ...
```
Done

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metrics/HoodieMetrics.java:
```diff
@@ -48,6 +49,7 @@
   private Timer deltaCommitTimer = null;
   private Timer finalizeTimer = null;
   private Timer compactionTimer = null;
+  private Timer clusteringTimer = null;
```
This is specific to clustering operation and only used in WriteClient#cluster. 
We could have a different timer for insertOverwrite and replace commands if needed. I like this approach because replace can mean multiple things. Let me know if you think common timer makes more sense.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/PartitionAwareScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,114 @@
+        return Option.empty();
+      }
+
+      long maxClusteringGroups = getWriteConfig().getClusteringMaxNumGroups();
```
Fixed.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/PartitionAwareScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,114 @@
+        maxClusteringGroups = MAX_CLUSTERING_GROUPS_STRATEGY;
+      }
+
+      List<HoodieClusteringGroup> clusteringGroups = partitionPaths.stream()
```
Updated.

satishkotha commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/schedule/strategy/SparkBoundedDayBasedScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,116 @@
+        .collect(Collectors.toList());
+  }
+
+  private int getNumberOfGroups(long groupSize, long targetFileSize) {
```
yes, this is calculating number of FileGroups. I updated the name to be explicit

satishkotha commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/schedule/strategy/SparkBoundedDayBasedScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,116 @@
+  @Override
+  protected List<String> filterPartitionPaths(List<String> partitionPaths) {
+    int targetPartitionsForClustering = getWriteConfig().getTargetPartitionsForClustering();
+    return partitionPaths.stream().map(partition -> partition.replace("/", "-"))
```
Seems to work fine without replacing '/'.  So I removed it. Btw, compaction has this code, so I'm not sure if there are special cases where this replace is needed. It should be easy to add if we run into any issue. So I removed it.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,149 @@
+  /**
+   * Return file slices eligible for clustering. FileIds in
+   * 1) pending clustering/compaction
+   * 2) Larger than clustering target file size
```
Moved 2 to strategy specific implementation.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ScheduleClusteringStrategy.java:
```diff
@@ -0,0 +1,149 @@
+  private final HoodieEngineContext engineContext;
+  private final HoodieWriteConfig writeConfig;
+
+  public ScheduleClusteringStrategy(HoodieTable table, HoodieEngineContext engineContext, HoodieWriteConfig writeConfig) {
```
I followed compaction naming pattern. But agree schedule has different meaning. I changed it to ClusteringPlanStrategy. WriteClient methods are still scheduleClustering(String instant), let me know if you think this needs to change also.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/AbstractBulkInsertHelper.java:
```diff
@@ -27,8 +27,21 @@
+                               boolean performDedupe,
+                               Option<BulkInsertPartitioner<T>> userDefinedBulkInsertPartitioner,
+                               boolean addMetadataFields,
+                               int numOutputGroups);
```
Looks like I already changed subclasses, but overlooked base class. Thanks for catching this.

satishkotha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/strategy/LogFileSizeBasedCompactionStrategy.java:
```diff
@@ -40,21 +36,6 @@
 public class LogFileSizeBasedCompactionStrategy extends BoundedIOCompactionStrategy
     implements Comparator<HoodieCompactionOperation> {
 
-  private static final String TOTAL_LOG_FILE_SIZE = "TOTAL_LOG_FILE_SIZE";
```
I refactored it based on other review feedback. There are unit tests for this. So I am reasonably confident theres no errors. I'll keep such changes for another PR next time. Thanks

satishkotha commented:
> @satishkotha
> 
> > After clustering all new records have a new commit time. I'm trying to see if it's possible to preserve commit_time from original file to support incremental queries.
> 
> this would be good to do. Can we file a tracking JIRA for this? we can keep under the same umbrella with 0.8.0 fix version?

@vinothchandar I created https://issues.apache.org/jira/browse/HUDI-1468 as suggesed.

satishkotha commented:
> This is great, thanks @satishkotha !
> 
> I have completed a first pass. Don't have major concerns. May be we can work through these initial comments, as I complete the remainder.
> 
> On follow ups
> 
> * IIUC inline clustering should work as-is from datasource/deltastreamer paths with this change, by passing necessary configs. We should create two JIRA one each for support for async clustering via datasource and deltastreamer?
> * Can you share how much testing on a production environment has been done for this.

@vinothchandar 

* Looks like @lw309637554 created https://issues.apache.org/jira/browse/HUDI-1399 for async clustering and there's some good discussion there.

* I did some basic testing in staging environment mostly with inline clustering. I have another PR for test suite changes to validate async clustering via calling writeClient APIs. I hope to get more production scale tests over next week.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+      "org.apache.hudi.client.clustering.plan.strategy.SparkBoundedDayBasedClusteringPlanStrategy";
+
+  // Config to provide a strategy class to execute a ClusteringPlan. Class has to be subclass of RunClusteringStrategy
+  public static final String RUN_CLUSTERING_STRATEGY_CLASS = "hoodie.clustering.run.strategy.class";
```
may be call it `execution`.  `hoodie.clustering.execution.strategy.class` ?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  private static final String DEFAULT_INLINE_CLUSTERING = "false";
+
+  // Config to control frequency of clustering
+  public static final String INLINE_CLUSTERING_NUM_COMMIT_PROP = "hoodie.clustering.inline.num.commits";
```
max.commits?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  public static final String INLINE_CLUSTERING_NUM_COMMIT_PROP = "hoodie.clustering.inline.num.commits";
+  private static final String DEFAULT_INLINE_CLUSTERING_NUM_COMMITS = "4";
+
+  // Number of partitions to scan to create ClusteringPlan.
```
replace `scan` with `list`? 

should we also do it based on last N commits? It helps for tables that receive data across partitions? This is probably some follow on work we can do to add a new plan strategy for this. Lets add a JIRA?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  private static final String DEFAULT_INLINE_CLUSTERING_NUM_COMMITS = "4";
+
+  // Number of partitions to scan to create ClusteringPlan.
+  public static final String CLUSTERING_TARGET_PARTITIONS = "hoodie.clustering.target.partitions";
```
`hoodie.clustering.plan.strategy.daybased.lookback.partitions` or something ties this to a specific strategy and also captures that this look at N partitions from now.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  public static final String CLUSTERING_TARGET_PARTITIONS = "hoodie.clustering.target.partitions";
+  public static final String DEFAULT_CLUSTERING_TARGET_PARTITIONS = String.valueOf(2);
+
+  // Each clustering operation can create multiple groups. Total amount of data processed by clustering operation
```
something to think about. should this be strategy specific too?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  // Each clustering operation can create multiple groups. Total amount of data processed by clustering operation
+  // is defined by below two properties (CLUSTERING_MAX_BYTES_IN_GROUP * CLUSTERING_MAX_NUM_GROUPS).
+  // Max amount of data to be included in one group
+  public static final String CLUSTERING_MAX_BYTES_IN_GROUP = "hoodie.clustering.max.bytes.group";
```
may be `hoodie.clustering.max.bytes.per.group` ?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieClusteringConfig.java:
```diff
@@ -0,0 +1,160 @@
+  public static final String CLUSTERING_MAX_NUM_GROUPS = "hoodie.clustering.max.num.groups";
+  public static final String DEFAULT_CLUSTERING_MAX_NUM_GROUPS = "30";
+
+  // Each group can produce 'N' (CLUSTERING_MAX_GROUP_SIZE/CLUSTERING_TARGET_FILE_SIZE) output file groups.
```
strategy specific?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/BaseCreateClusteringPlanActionExecutor.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Collections;
+import java.util.Map;
+
+public abstract class BaseCreateClusteringPlanActionExecutor<T extends HoodieRecordPayload, I, K, O> extends BaseActionExecutor<T, I, K, O, Option<HoodieClusteringPlan>> {
```
drop the `Create`?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ClusteringPlanStrategy.java:
```diff
@@ -0,0 +1,146 @@
+  /**
+   * Return file slices eligible for clustering. FileIds in
+   * 1) pending clustering/compaction
+   * 2) Larger than clustering target file size
```
remove this?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ClusteringPlanStrategy.java:
```diff
@@ -0,0 +1,146 @@
+    return slices.stream().map(slice -> new HoodieSliceInfo().newBuilder()
+        .setPartitionPath(slice.getPartitionPath())
+        .setFileId(slice.getFileId())
+        .setDataFilePath(slice.getBaseFile().map(BaseFile::getPath).orElse(""))
```
EMPTY_STRING?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/ClusteringPlanStrategy.java:
```diff
@@ -0,0 +1,146 @@
+  /**
+   * Transform {@link FileSlice} to {@link HoodieSliceInfo}.
+   */
+  protected List<HoodieSliceInfo> getFileSliceInfo(List<FileSlice> slices) {
```
worth thinking about making this a static helper?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/cluster/strategy/PartitionAwareClusteringPlanStrategy.java:
```diff
@@ -0,0 +1,108 @@
+/**
+ * Scheduling strategy with restriction that clustering groups can only contain files from same partition.
+ */
+public abstract class PartitionAwareClusteringPlanStrategy<T extends HoodieRecordPayload,I,K,O> extends ClusteringPlanStrategy<T,I,K,O> {
```
same something like DayBased? to indicate date partitioning?

vinothchandar replied:
also rename the subclass to indicate that it prefers the recent partitions?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/plan/strategy/SparkBoundedDayBasedClusteringPlanStrategy.java:
```diff
@@ -0,0 +1,124 @@
+  protected Stream<FileSlice> getFileSlicesEligibleForClustering(final String partition) {
+    return super.getFileSlicesEligibleForClustering(partition)
+        // files that have basefile size larger than clustering target file size are not eligible (Note that compaction can merge any updates)
+        .filter(slice -> slice.getBaseFile().map(HoodieBaseFile::getFileSize).orElse(0L) < getWriteConfig().getClusteringTargetFileMaxBytes());
```
Something like `hoodie.clustering.small.file.limit`

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDCustomColumnsSortPartitioner.java:
```diff
@@ -0,0 +1,66 @@
+    final String[] sortColumns = this.sortColumnNames;
+    final String schemaStr = this.schemaString;
+    return records.sortBy(record -> {
+      Schema schema = new Schema.Parser().parse(schemaStr);
```
we can add `SerializableSchema` (we do this for Hadoop Config object) ?

vinothchandar replied:
See `SerialiableConfiguration` for reference

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,230 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class SparkRunClusteringCommitActionExecutor<T extends HoodieRecordPayload<T>>
```
SparkClusteringCommitActionExecutor?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/cluster/SparkRunClusteringCommitActionExecutor.java:
```diff
@@ -0,0 +1,230 @@
+
+    JavaSparkContext engineContext = HoodieSparkEngineContext.getSparkContext(context);
+    // run clustering for each group async and collect WriteStatus
+    JavaRDD<WriteStatus> writeStatusRDD = clusteringPlan.getInputGroups().stream()
```
this is very cool!

vinothchandar reviewed (approved):
LGTM otherwise. Please merge when ready

vinothchandar commented:
Renaming suggestions

`SparkBoundedDayBasedClusteringPlanStrategy` -> `SparkRecentDaysClusteringPlanStrategy`
`RunClusteringStrategy` -> `ClusteringExecutionStrategy` 
`SparkBulkInsertBasedRunClusteringStrategy` -> `SparkSortAndSizeExecutionStrategy`

lw309637554 commented:
> > This is great, thanks @satishkotha !
> > I have completed a first pass. Don't have major concerns. May be we can work through these initial comments, as I complete the remainder.
> > On follow ups
> > 
> > * IIUC inline clustering should work as-is from datasource/deltastreamer paths with this change, by passing necessary configs. We should create two JIRA one each for support for async clustering via datasource and deltastreamer?
> > * Can you share how much testing on a production environment has been done for this.
> 
> @vinothchandar
> 
> * Looks like @lw309637554 created https://issues.apache.org/jira/browse/HUDI-1399 for async clustering and there's some good discussion there.
> * I did some basic testing in staging environment mostly with inline clustering. I have another PR for test suite changes to validate async clustering via calling writeClient APIs. I hope to get more production scale tests over next week.

@satishkotha  i had added inline clustering unit tests for spark datasource and deltastreamer in my local branch .  When this pr merged , i can open pull request. Then base on the unit tests , will land the independent clustering spark job , then async clustering  via datasource and via deltastreamer.   cc @vinothchandar

satishkotha commented:
Thanks @vinothchandar @n3nash @lw309637554 for comments. I merged this PR.

vinothchandar commented:
This is game changing. Thanks @satishkotha !

lw309637554 commented:
> Thanks @vinothchandar @n3nash @lw309637554 for comments. I merged this PR.

Great  ，then can do more interesting things based on this PR
